File: src/main/java/ru/citeck/ecos/notifications/domain/template/handler/NotificationTemplateModuleHandler.java
package ru.citeck.ecos.notifications.domain.template.handler;

import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.stereotype.Component;
import ru.citeck.ecos.apps.module.controller.type.binary.BinModule;
import ru.citeck.ecos.apps.module.handler.EcosModuleHandler;
import ru.citeck.ecos.apps.module.handler.ModuleMeta;
import ru.citeck.ecos.apps.module.handler.ModuleWithMeta;
import ru.citeck.ecos.commons.data.MLText;
import ru.citeck.ecos.commons.data.ObjectData;
import ru.citeck.ecos.commons.io.file.mem.EcosMemDir;
import ru.citeck.ecos.commons.utils.ZipUtils;
import ru.citeck.ecos.notifications.domain.template.dto.NotificationTemplateDto;
import ru.citeck.ecos.notifications.domain.template.dto.TemplateDataDto;
import ru.citeck.ecos.notifications.domain.template.service.NotificationTemplateService;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Component
@RequiredArgsConstructor
public class NotificationTemplateModuleHandler implements EcosModuleHandler<BinModule> {

    private final NotificationTemplateService templateService;

    private static final Pattern LANG_KEY_PATTERN = Pattern.compile(".*_(\\w+).*");
    private static final String DEFAULT_LANG_KEY = "en";

    @Override
    public void deployModule(@NotNull BinModule module) {
        templateService.update(toDto(module));
    }

    private NotificationTemplateDto toDto(BinModule module) {
        NotificationTemplateDto dto = new NotificationTemplateDto();

        ObjectData meta = module.getMeta();

        dto.setId(meta.get("id").asText());
        dto.setTitle(meta.get("title", MLText.class));

        EcosMemDir memDir = ZipUtils.extractZip(module.getData());
        dto.setData(getTemplateDataFromMemDir(memDir));

        return dto;
    }

    private Map<String, TemplateDataDto> getTemplateDataFromMemDir(EcosMemDir memDir) {
        Map<String, TemplateDataDto> templateData = new HashMap<>();

        memDir.getChildren().forEach(file -> {
            String langKey = getLangKeyFromFileName(file.getName());

            TemplateDataDto dataDto = new TemplateDataDto();
            dataDto.setName(file.getName());
            dataDto.setData(file.readAsBytes());

            templateData.put(langKey, dataDto);
        });

        return templateData;
    }

    private String getLangKeyFromFileName(String fileName) {
        Matcher matcher = LANG_KEY_PATTERN.matcher(fileName);
        if (matcher.find()) {
            return matcher.group(1);
        }
        return DEFAULT_LANG_KEY;
    }

    @NotNull
    @Override
    public ModuleWithMeta<BinModule> getModuleMeta(@NotNull BinModule module) {
        NotificationTemplateDto dto = toDto(module);
        return new ModuleWithMeta<>(module, new ModuleMeta(dto.getId(), Collections.emptyList()));
    }

    @NotNull
    @Override
    public String getModuleType() {
        return "notification/template";
    }

    @Override
    public void listenChanges(@NotNull Consumer<BinModule> consumer) {
    }

    @Nullable
    @Override
    public ModuleWithMeta<BinModule> prepareToDeploy(@NotNull BinModule module) {
        return getModuleMeta(module);
    }


}

File: src/main/java/ru/citeck/ecos/notifications/domain/template/handler/NotificationTemplateModuleHandler.kt
package ru.citeck.ecos.notifications.domain.template.handler

import mu.KotlinLogging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import ru.citeck.ecos.apps.module.controller.type.binary.BinModule
import ru.citeck.ecos.apps.module.handler.EcosModuleHandler
import ru.citeck.ecos.apps.module.handler.ModuleMeta
import ru.citeck.ecos.apps.module.handler.ModuleWithMeta
import ru.citeck.ecos.commons.data.MLText
import ru.citeck.ecos.commons.io.file.EcosFile
import ru.citeck.ecos.commons.io.file.mem.EcosMemDir
import ru.citeck.ecos.commons.utils.ZipUtils.extractZip
import ru.citeck.ecos.notifications.domain.template.dto.NotificationTemplateWithMeta
import ru.citeck.ecos.notifications.domain.template.dto.TemplateDataDto
import ru.citeck.ecos.notifications.domain.template.getLangKeyFromFileName
import ru.citeck.ecos.notifications.domain.template.service.NotificationTemplateService
import java.util.*
import java.util.function.Consumer

@Component
class NotificationTemplateModuleHandler : EcosModuleHandler<BinModule> {

    companion object {
        private val log = KotlinLogging.logger {}
    }

    @Autowired
    private lateinit var templateService: NotificationTemplateService

    override fun deployModule(module: BinModule) {
        templateService.update(toDto(module))
    }

    private fun toDto(module: BinModule): NotificationTemplateWithMeta {
        val meta = module.meta

        val dto = NotificationTemplateWithMeta(meta.get("id").asText())
        dto.notificationTitle = meta.get("notificationTitle", MLText::class.java)
        dto.name = meta.get("name").asText()
        val memDir = extractZip(module.data)
        dto.templateData = getTemplateDataFromMemDir(memDir)
        dto.model = meta.get("model").asMap(String::class.java, String::class.java)

        log.debug("Deploy new template module: $dto")
        if (log.isDebugEnabled) {
            memDir.getChildren().forEach {
                val name = it.getName()
                val content = it.readAsString()
                log.debug("name: $name content: \n$content")
            }
        }

        return dto
    }

    private fun getTemplateDataFromMemDir(memDir: EcosMemDir): Map<String, TemplateDataDto> {
        val templateData: MutableMap<String, TemplateDataDto> = HashMap()
        memDir.getChildren().forEach(Consumer { file: EcosFile ->
            val langKey = getLangKeyFromFileName(file.getName())
            val dataDto = TemplateDataDto(file.getName(), file.readAsBytes())
            templateData[langKey] = dataDto
        })
        return templateData
    }

    override fun getModuleMeta(module: BinModule): ModuleWithMeta<BinModule> {
        val dto = toDto(module)
        return ModuleWithMeta(module, ModuleMeta(dto.id, emptyList()))
    }

    override fun getModuleType(): String {
        return "notification/template"
    }

    override fun listenChanges(listener: Consumer<BinModule>) {
        //TODO: implement
    }

    override fun prepareToDeploy(module: BinModule): ModuleWithMeta<BinModule>? {
        return getModuleMeta(module)
    }
}
